### PR TITLE
Fix handling of sessions and kernels restarts

### LINKED ISSUE #1464: Pressing "Restart and run all" randomly causes silent pyodide kernel crash
## Description

Pressing "Restart and run all" randomly causes silent pyodide kernel crashes as seen on the following screen recording:

[jupyterlite_pyodide_crash.webm](https://github.com/user-attachments/assets/1691367d-29cc-4fc9-949e-f5ddb560e942)

## Reproduce

1. Go to https://jupyter.org/try-jupyter/lab/ in chrome or firefox
2. Create a new notebook with a single cell that imports a builtin library such as `six`
3. Execute the cell, in general there is no problem at this point.
4. Press the "Restart and run all cells" but a few times
5. At some point, the pyodide kernel crashes
6. The browser dev console displays a message such as:

```
Trying to send message on removed socket for kernel a7355f7a-283a-4cdb-a37c-ee379fb1bfc7
```

## Context

- JupyterLite version: I can reproduce on todays' dev version but also on the version currently hosted at https://jupyter.org/try-jupyter/lab/)
- Operating System and version: up to date macOS (e.g. 14.5) and Ubuntu LTS
- Browser and version: I tried on several browsers (up to date chrome and firefox, e.g. 129.0.2).

<details><summary>Browser Output</summary>
<pre>
Kernel: restarting (a7355f7a-283a-4cdb-a37c-ee379fb1bfc7) [default.js:1370:24](webpack://_JUPYTERLAB.CORE_OUTPUT/node_modules/@jupyterlab/services/lib/kernel/default.js)
Pyodide contents will be synced with Jupyter Contents [index.js:60:28](webpack://jupyterlite/pyodide-kernel-extension/lib/index.js)
Connection lost, reconnecting in 0 seconds. [default.js:1325:20](webpack://_JUPYTERLAB.CORE_OUTPUT/node_modules/@jupyterlab/services/lib/kernel/default.js)
Starting WebSocket: wss://jupyter.org/try-jupyter/api/kernels/a7355f7a-283a-4cdb-a37c-ee379fb1bfc7 [default.js:69:20](webpack://_JUPYTERLAB.CORE_OUTPUT/node_modules/@jupyterlab/services/lib/kernel/default.js)
TypeError: this is undefined
    isReady notebooklspadapter.js:187
    y utils.js:22
    y utils.js:20
    onKernelChanged notebooklspadapter.js:149
    c index.es6.js:555
    emit index.es6.js:513
    emit index.es6.js:112
    restartKernel sessioncontext.js:366
    restart sessioncontext.js:882
    execute index.js:1801
    execute index.es6.js:365
    onClick toolbar.js:1043
    o toolbar.js:667
    React 11
[notebooklspadapter.js:163:20](webpack://_JUPYTERLAB.CORE_OUTPUT/node_modules/@jupyterlab/notebook/lib/notebooklspadapter.js)
Loading micropip, packaging [pyodide.asm.js:10:93500](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
Loaded micropip, packaging [pyodide.asm.js:10:93796](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
History was unable to be retrieved [history.js:260:20](webpack://_JUPYTERLAB.CORE_OUTPUT/node_modules/@jupyterlab/notebook/lib/history.js)
Loading openssl, ssl [pyodide.asm.js:10:93500](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
Loaded openssl, ssl [pyodide.asm.js:10:93796](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
Loading sqlite3 [pyodide.asm.js:10:93500](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
Loaded sqlite3 [pyodide.asm.js:10:93796](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
Loading traitlets [pyodide.asm.js:10:93500](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
Loaded traitlets [pyodide.asm.js:10:93796](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
traitlets already loaded from default channel [pyodide.asm.js:10:93166](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
sqlite3 already loaded from default channel [pyodide.asm.js:10:93166](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
Loading Pygments, asttokens, decorator, executing, ipython, matplotlib-inline, prompt_toolkit, pure_eval, six, stack_data, wcwidth [pyodide.asm.js:10:93500](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
Loaded Pygments, asttokens, decorator, executing, ipython, matplotlib-inline, prompt_toolkit, pure_eval, six, stack_data, wcwidth [pyodide.asm.js:10:93796](https://cdn.jsdelivr.net/pyodide/v0.26.2/full/pyodide.asm.js)
Failed to fetch ipywidgets through the "jupyter.widget.control" comm channel, fallback to fetching individual model state. Reason: Control comm did not respond in time [327.68dbf8491690b3aff1e7.js:1:5783](https://jupyter.org/try-jupyter/extensions/@jupyter-widgets/jupyterlab-manager/static/327.68dbf8491690b3aff1e7.js?v=68dbf8491690b3aff1e7)
Trying to send message on removed socket for kernel a7355f7a-283a-4cdb-a37c-ee379fb1bfc7 3 [kernels.js:108:24](webpack://_JUPYTERLAB.CORE_OUTPUT/packages/kernel/lib/kernels.js)
</pre>
</details>

### PR DESCRIPTION
**Pull request by jtpio**
## References

Fixes #1464

## Code changes

- [x] Bump `mock-socket`
- [x] Fix session handling logic on kernel restarts
- [x] Add UI test

## User-facing changes

[jupyterlite-kernel-restarts.webm](https://github.com/user-attachments/assets/5050739c-1f8d-452c-8e33-ed02c494ad88)

## Backwards-incompatible changes

None

### CODE REVIEW

File: packages/session/src/sessions.ts [hunk 1]
```diff
@@ -170,7 +170,9 @@ export class Sessions implements ISessions {
     const runningKernel = await this._kernels.get(kernelId);
     if (runningKernel) {
       runningKernel.disposed.connect(() => {
-        this.shutdown(sessionId);
+        // eslint-disable-next-line no-console
+        console.log('Session', sessionId, 'shut down');
+        // this.shutdown(sessionId);
```
Comment (lesteve) [inline packages/session/src/sessions.ts:175]: Just a note for our later selves, removing `this.shutdown(sessionId)` seems to be one way to make the "No Kernel on restart" issue go away.

Comment (jtpio) [reply packages/session/src/sessions.ts]: Yes, but we should also handle shutting down via the kernel list:

image

I pushed a change in https://github.com/jupyterlite/jupyterlite/pull/1538/commits/f6beaa7cb3a2560626ccd1b1bb8614511e9cc352 to handle this.

**Comment by jtpio** [comment]
cc @ogrisel @lesteve if you would like to give it a try: https://jupyterlite--1538.org.readthedocs.build/en/1538/_static/lab/index.html

There may be other issues related to kernel and session management, but should hopefully already improve the case of kernel restarts.

**Comment by jtpio** [comment]
>  * [ ]  Add UI test

We should able to put this more under tests. Currently there is a test to make sure a kernel can be shut down and re-opened, but there could be another one to perform multiple restarts in a row, potentially with the JS kernel so it's faster to start:

https://github.com/jupyterlite/jupyterlite/blob/e0d986fec32ef5f55d4e15db30061fa501509ec3/ui-tests/test/kernels.spec.ts#L48-L49

**Comment by lesteve** [comment]
Super nice, thanks!

I tried with a Pyodide kernel and did not manage to trigger the issue with your [preview link](https://jupyterlite--1538.org.readthedocs.build/en/1538/_static/lab/index.html).

I can easily trigger the issue with a Pyodide kernel in https://jupyter.org/try-jupyter/lab/.

I would be motivated to try to add a UI test mostly as an exercise to get more familiar with Jupyterlite development, but it would take me a bit of time I think, so if you want to add the test yourself it's fine as well, do what you think is best ...

ogrisel is AFK at the moment I think.

**Comment by jtpio** [comment]
> I would be motivated to try to add a UI test mostly as an exercise to get more familiar with Jupyterlite development, but it would take me a bit of time I think, so if you want to add the test yourself it's fine as well, do what you think is best ...

Thanks for proposing to help!

Since I have the dev setup locally, maybe I'll go ahead this time and add a test in this PR directly.

But there are more UI tests we could add in general (for example https://github.com/jupyterlite/jupyterlite/issues/1027), if you are interested :)

**Comment by jtpio** [comment]
UI test seems to be passing on CI (video of the local run):

[e5a2758048efc7b9d2b3f268f6999649feb73d5a.webm](https://github.com/user-attachments/assets/dbd4af88-161e-485d-b51c-f8e59723580d)

**Comment by jtpio** [comment]
> Just for completeness I tried to see whether that would fix #1359 and #1527 but it doesn't. I guess they happen because of memory limitations inside the browser.

Out of curiosity, how many kernels did you start before it started to become an issue? 

Wondering if a way to mitigate that issue could be to inform users (in the UI) they could terminate some unused kernels if too many are running?

**Comment by jtpio** [comment]
OK cool. Let me know if you would like to do a bit more testing with this PR.

Otherwise, happy to get it in and release a new 0.5.0 pre-release, if it already helps fix #1464.

**Comment by ogrisel** [comment]
I confirm I can no longer reproduce #1464. Thanks so much @jtpio! It will make using jupyterlite much more reliable to teach python-based data science skills!

+1 for merging as is.

**Comment by lesteve** [comment]
I did some testing with a realistic notebook (that take roughly 1min30s to execute fully) and I was not able to reproduce the issue with 5 "restart and run all cells".

**Comment by jtpio** [comment]
Thanks @lesteve @ogrisel for testing!

Merging and releasing in `0.5.0a1` then :+1:

**Comment by lesteve** [comment]
Great to see this fixed :tada:, thanks a lot!

**Comment by jtpio** [comment]
Released: https://github.com/jupyterlite/jupyterlite/releases/tag/v0.5.0a1

**Pull request merged**